routes/api/activity.js

```diff
@@ -0,0 +1,1167 @@
+        LEFT JOIN ( \
+            SELECT \
+                a.id AS "activityId", \
+                jpr.*, \
+                g.name AS "groupName", \
+                t.title AS "topicTitle" \
+                FROM "Activities" a \
+                JOIN jsonb_to_record(a.data) AS jpr("@type" text, "level" text, "groupId" text, "topicId" text, "createdAt" text, "updatedAt" text, "deletedAt" text) ON a.id = a.id \
+                JOIN "Groups" g ON g.id::text = data#>>\'{origin, groupId}\' \
+                JOIN "Topics" t ON t.id::text = data#>>\'{origin, topicId}\' \
+                WHERE a.data#>>\'{origin, @type}\' = \'TopicMemberGroup\' \
+        ) tmgorig ON a.data#>>\'{origin, @type}\' = \'TopicMemberGroup\' AND tmuorig."activityId" = a.id \
+        LEFT JOIN ( \
+            SELECT \
+                a.id AS "activityId", \
+                jpr.*, \
+                u.name AS "userName", \
+                g.name AS "groupName" \
+                FROM "Activities" a \
+                JOIN jsonb_to_record(a.data) AS jpr("@type" text, "level" text, "userId" text, "groupId" text, "createdAt" text, "updatedAt" text, "deletedAt" text) ON a.id = a.id \
+                JOIN "Users" u ON u.id::text = data#>>\'{origin, userId}\' \
+                JOIN "Groups" g ON g.id::text = data#>>\'{origin, groupId}\' \
+                WHERE a.data#>>\'{origin, @type}\' = \'GroupMember\' \
+        ) gmuorig ON a.data#>>\'{origin, @type}\' = \'GroupMember\' AND tmuorig."activityId" = a.id \
+        LEFT JOIN \
+            "Groups" targg ON a.data#>>\'{target, @type}\' = \'Group\' AND targg.id::text = a.data#>>\'{target, id}\' \
+        LEFT JOIN \
+            "Topics" targt ON a.data#>>\'{target, @type}\' = \'Topic\' AND targt.id::text = a.data#>>\'{target, id}\' \
+        LEFT JOIN ( \
+            SELECT \
+                a.id AS "activityId", \
+                c.*, \
+                tc."topicId" \
+                FROM "Activities" a \
+                JOIN "Comments" c ON a.data#>>\'{target, @type}\' = \'Comment\' AND c."id"::text = a.data#>>\'{target, id}\' \
+                JOIN "TopicComments" tc ON tc."commentId" = c.id \
+        ) targtc ON a.data#>>\'{target, @type}\' = \'Comment\' AND targtc."id"::text = a.data#>>\'{target, id}\' \
+    ';
+
+    var buildActivityFeedIncludeString = function (req, visibility) {
+        var queryInclude = req.query.include || [];
+        
+        if (queryInclude && !Array.isArray(queryInclude)) {
+            queryInclude = [queryInclude];
+        }
+        var allowedIncludeAuth = ['userTopics', 'userGroups', 'user', 'self'];
+        var allowedInclude = ['publicTopics', 'publicGroups'];
+        
+        var include = queryInclude.filter(function (item, key, input) {
+            if (visibility && visibility === 'public') {
+                return allowedInclude.indexOf(item) > -1 && (input.indexOf(item) === key);
```

Comment:
OK. It can stay as is.

For future reference - we could use Set ("is allowed, push to set") which guarantees uniqueness. But that's just a different way of solving the same problem. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set